plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -37,6 +37,30 @@ export async function eachBatchIngestion(payload: EachBatchPayload, queue: Inges
     await eachBatch(payload, queue, eachMessageIngestion, groupIntoBatchesIngestion, 'ingestion')
 }
 
+export async function eachBatchParallelIngestion(payload: EachBatchPayload, queue: IngestionConsumer): Promise<void> {
+    function groupByTeamDistinctId(kafkaMessages: KafkaMessage[]): PipelineEvent[][] {
+        const batches: Map<string, PipelineEvent[]> = new Map()
+        for (const message of kafkaMessages) {
+            const pluginEvent = formPipelineEvent(message)
+            const key = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
+            const siblings = batches.get(key)
+            if (siblings) {
+                siblings.push(pluginEvent)
+            } else {
+                batches.set(key, [pluginEvent])
+            }
+        }
+
+        return Array.from(batches.values())
+    }
+
+    async function eachMessage(event: PipelineEvent, queue: IngestionConsumer): Promise<void> {
+        await ingestEvent(queue.pluginsServer, queue.workerMethods, event)
+    }
+
+    await eachBatchParallel(payload, queue, eachMessage, groupByTeamDistinctId, 'ingestion')
```

Comment:
Is is possible that we don't need to do this higher functional stuff and simply apply `groupByTeamDistinctId` here, and kick of the processing on them? I dislike all the passing around of functions tbh, I get so confused.

Reply:
Agreed. I kept the same patterm to make reasoning about the change easier now, but I'll try to consolidate stuff together when I rebuild overflow in it.